torch/lib/c10d/logger.cpp

```diff
@@ -85,7 +91,148 @@ void Logger::set_construction_data_and_log(
   LogPyTorchDDPUsage(*ddp_logging_data_);
 }
 
+void Logger::calculate_avg_cpu_time(
+    int64_t& avg_time,
+    int64_t cpu_start_time,
+    int64_t cpu_end_time) {
+  // If cpu_end_time is not recorded in this iteration,
+  // avg_time will return invalid value.
+  // For some cases like DDP runs on non-sync mode, backward compute
+  // end time can not be recorded in this iteration and thus can not
+  // calculate the valid avg_time.
+  // In this case, skip calculating the avg_time and return.
+  TORCH_CHECK(num_iterations_stats_recorded_ > 0);
+  if (cpu_end_time < cpu_start_time) {
+    return;
+  }
+  avg_time = ((cpu_end_time - cpu_start_time) +
+              avg_time * (num_iterations_stats_recorded_ - 1)) /
+      num_iterations_stats_recorded_;
+}
+
+#ifdef USE_CUDA
+void Logger::calculate_avg_gpu_time(
+    int64_t& avg_time,
+    at::cuda::CUDAEvent& gpu_start,
+    at::cuda::CUDAEvent& gpu_end) {
+  TORCH_CHECK(num_iterations_stats_recorded_ > 0);
+  float milliseconds = gpu_start.elapsed_time(gpu_end);
+  // If gpu_end is not recorded in this iteration,
+  // milliseconds will have invalid value.
+  // For some cases like DDP runs on non-sync mode,
+  // gpu_end can not be recorded in this iteration and thus can not
+  // calculate the valid avg_time.
+  // In this case, skip calculating the avg_time and return.
+  if (milliseconds < 0) {
+    return;
+  }
+  avg_time = (int(milliseconds * kMilliSecondToNanosSecond) +
```

Comment:
nit: curious, will it be more efficient and accurate if we store the total time, and then calculate the `avg_time` on demand when needed?

Reply:
I guess the efficiency improvement here will be very small, calculating avg_time at sampling iterations should be cheap?